### ISSUE TITLE
Communication Architecture

### ISSUE THREAD

**Issue by metaloph1l** [metaloph1l/pacman#4]
Design of an extensible communication system between the client and the server. This does NOT include implementations for all the actions but rather a general architecture of the necessary work flows. Sub issues have to be opened for the implementations.

Does include at least:
Server has to be able to send updates to the client when the state of the game changes, e.g. mapchange, color change, movement, game over.
Client has to be able to send commands to the server that the user directs to the program. Basic actions are up, down, left, right.

Other actions include disconnect, connect and polling of the current game state.
Pause is not needed, because the game should never pause.

Server Package: at.ac.foop.pacman.gameserver.communication
Client Package: at.ac.foop.pacman.client.communication

**Comment by lanoxx** [comment]
Also fuer die Kommunikation vom Server zum Client brauchen wir 
eigentlich nur das Observer Pattern verwenden. Mit RMI muessten wir dann 
nur beim Server ein Callback Objekt fuer jeden Client Registrieren. Und 
wenn wir Sockets verwenden ist eh immer eine Verbindung offen ueber die 
der Server direkt den Befehl senden kann.

Ich wuerde fast sagen das Sockets ausreichen fuer die Aufgabenstellung. 
Das Protokoll ergibt sich quasi von selbst aus der Aufgabenstellung und 
wenn wir es einmal definiert haben dann aendert es sich auch nicht mehr.

Die Frage ist halt ob es mit RMI evtl. weniger Arbeit waere. Zu Space 
Based Middleware kann ich nichts sagen, habe ich bisher noch nicht 
verwendet und hoere gerade erst die Vorlesung dazu.

**Comment by metaloph1l** [comment]
Ja, stimmt. Die Kommunikation kann eigentlich ganz einfach ablaufen:
- Clients senden ihre Befehle an den Server
- Server updated nach jeder Aktion das Spiel im Modell und schickt die Änderungen an die Clients aus. Hier geht es vor allem darum eine Möglichkeit zu finden nur kleinstmögliche Änderungen an die Clients zu schicken, da es sonst problematisch werden kann, wenn wir immer das gesamte Spiel verschicken.
- Clients updaten die Repräsentation wenn eine Änderung eintrifft.

In diesem Fall bin ich mir nicht einmal sicher ob wir hier das Observer-Pattern brauchen. Es handelt sich eher um einen verteilten Observer.

Bei der Kommunikation hast du recht. Sockets würden sicher ausreichen (so wie eigentlich überall), allerdings ist RMI oder XVSM sicher weniger aufwendig, da wir die ganze Kommunikationsschicht wegen der Middleware nicht programmieren müssen. Über RMI und XVSM kann man ziemlich einfach Objekte verschicken, bei Sockets muss man dabei bereits diese Funktionalität nachbauen.

Ich denke nicht, dass wir hier Sockets verwenden sollten, da es mit den anderen Technologien die Kommunikation einfacher und unaufwendiger ist und die low-level Kommunikation nicht im Fokus der LVA steht.

**Comment by lanoxx** [comment]
ok dann ist die Frage ob wir RMI oder dieses XVSM verwenden, wie gesagt, 
mir Space Based middle ware habe ich keine Erfahrung, also wuerde ich 
fuer RMI stimmen, ausser es gibt einen entscheidenden Vorteil bei XVSM?

**Comment by metaloph1l** [comment]
Ich denke, dass für die Aufgabenstellung RMI locker ausreicht.

Mit XVSM wollte ich nur eine andere Option anbieten. Damit kenne ich mich eigentlich ziemlich gut aus und die Spaces sind generell sehr einfach und intuitiv gehalten, d.h. wenn ihr generell interessiert daran seid, braucht ihr euch davon nicht abschrecken zu lassen, weil ihr bei konkreten Dingen immer mich fragen könnt und ich ja auch mit arbeite. Ich hätte auch bereits einige Utility-Klassen aus einem anderen Projekt die wir wieder verwenden könnten. Mit RMI kenne ich mich aber auch gut aus.

**Comment by spookyTU** [comment]
Also mir ist wie gesagt egal ob es XVSM oder RMI wird. 
Zu Sockets: Das Protokoll ergibt sich eigentlich - wie bereits gesagt - aus der Angabe. Ob wir tatsaechlich Objekte uebertragen muessen, kann ich jetzt noch nicht sagen (aber der Aufwand wuerde wahrscheinlich geringer werden). Wir haben ja eigentlich nicht wirklich viele Notifications vom Server (evtl. neuer Spieler, Spieler wurde gefressen usw.). 

Das Spielfeld (bzw. den aktuellen Zustand) muessen wir aber denke ich jedes mal uebertragen (ich glaub auch das hier komplett am einfachsten sein sollte), aber das haengt von der konkreten Implementierung ab. Wenn wir einfach das Spielfeld als Matrix (wobei jedes Feld einen bestimmten Zustand hat) uebertragen und am Client dann entsprechend rendern, waere auch der Kommunikations-Overhead nicht so gross.

Per RMI oder XVSM sind wir halt insofern flexibler als wir einfach neue Methoden fuer einzelne Teile implementieren koennen und diese dann aufrufen. Wir brauchen uns dann nicht um ein eigentliches Protokoll zu kuemmern und dieses anzupassen (falls Aenderungen auftreten)

**Comment by metaloph1l** [comment]
Does anyone have any serious problems with using RMI? It seems like the optimal solution at the moment.

**Comment by lanoxx** [comment]
So whats left for this issue is to write the RMI registration code for the client and the server. I suggest to put that into the ui startup code. What do you think?
